crates/circuit_encodings/src/lib.rs

```diff
@@ -222,10 +245,9 @@ impl<
             previous_head: self.head, // unchanged
             previous_tail: old_tail,
             num_items: self.num_items,
-            round_function_execution_pairs: states,
         };
 
-        (old_tail, intermediate_info)
+        (round_states, intermediate_info)
     }
 
     pub fn pop_and_output_intermediate_data<
```

Comment:
If you want to make a large performance benefit overhaul, then it's better to focus on storing some intermediate information instead of using round function to update "head" when one pops an element from the queue

Reply:
Specifically, this change is aimed rather at reducing the use of RAM. And these particular simulators work relatively quickly (compared to memory queue simulators).

Still, the idea is good, I propose to think about it in the next pull requests